edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -86,45 +91,211 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
+            var moduleId = match.Groups["id2"].Success ? Option.Some<string>(match.Groups["id2"].Value) : Option.None<string>();
+
+            // For indirect clients the subscription change will be reported through edgeHub. That is a
+            // slower method to reconcile so we handle the direct case separately with a faster method
+            await moduleId.Filter(i => string.Equals(i, Constants.EdgeHubModuleId))
+                     .Match(
+                        _ => this.HandleNestedSubscriptionChanges(subscriptionList),
+                        () => this.HandleDirectSubscriptionChanges(deviceId, moduleId, subscriptionList));
+
+            return true;
+        }
+
+        async Task HandleDirectSubscriptionChanges(Option<string> deviceId, Option<string> moduleId, List<string> subscriptionList)
+        {
+            var identity = moduleId.Match(
+                               mod => deviceId.Match(
+                                         dev => this.identityProvider.Create(dev, mod),
+                                         () => throw new ArgumentException("Invalid topic structure for subscriptions - Module name matched but no Device name")),
+                               () => deviceId.Match(
+                                         dev => this.identityProvider.Create(dev),
+                                         () => throw new ArgumentException("Invalid topic structure for subscriptions - no Device name matched")));
+
+            var listener = default(IDeviceListener);
+            var maybeListener = await this.connectionRegistry.GetDeviceListenerAsync(identity, true);
+
+            if (maybeListener.HasValue)
+            {
+                listener = maybeListener.Expect(() => new Exception($"No device listener found for {identity.Id}"));
+
+                foreach (var subscriptionPattern in this.subscriptionPatterns)
+                {
+                    var subscribes = false;
+
+                    foreach (var subscription in subscriptionList)
+                    {
+                        var subscriptionMatch = Regex.Match(subscription, subscriptionPattern.Pattern);
+                        if (subscriptionMatch.Success)
+                        {
+                            var subscribedDevId = subscriptionMatch.Groups["id1"].Success ? Option.Some<string>(subscriptionMatch.Groups["id1"].Value) : Option.None<string>();
+                            var subscribedModId = subscriptionMatch.Groups["id2"].Success ? Option.Some<string>(subscriptionMatch.Groups["id2"].Value) : Option.None<string>();
+
+                            if (IsMatchingIds(subscribedDevId, subscribedModId, deviceId, moduleId))
+                            {
+                                subscribes = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    try
```

Comment:
try - catch could be moved inside AddOrRemoveSubscription

Reply:
moved into the method